actionpack/lib/action_dispatch/routing/route_set.rb

```diff
@@ -617,7 +621,7 @@ def current_controller
         def use_recall_for(key)
           if @recall[key] && (!@options.key?(key) || @options[key] == @recall[key])
             if !named_route_exists? || segment_keys.include?(key)
-              @options[key] = @recall.delete(key)
+              @options[key] = @recall[key]
```

Comment:
I agree, it raised some red flags with me when I first did it. Here's the specific commit:

https://github.com/schneems/rails/commit/9060b92f406b23ac793623de7c8c0eb6ebb8cd86

It's a fairly large savings in memory. Here's why I think it's okay

This method moves a key/value pair from recall to options

```ruby
def use_recall_for(key)
  if @recall[key] && (!@options.key?(key) || @options[key] == @recall[key])
    if !named_route_exists? || segment_keys.include?(key)
      @options[key] = @recall[key]
    end
  end
end
```

I changed it so that recall is preserved i.e. the value is copied not "moved." This method gets called 3 times for the keys `:controller`, `:action`, and `:id`:

```ruby
# This pulls :controller, :action, and :id out of the recall.
# The recall key is only used if there is no key in the options
# or if the key in the options is identical. If any of
# :controller, :action or :id is not found, don't pull any
# more keys from the recall.
def normalize_controller_action_id!
  use_recall_for(:controller) or return
  use_recall_for(:action) or return
  use_recall_for(:id)
end
```

Based on the comment every time `options` would have a key, it should be favored over recall, so it doesn't matter that recall also has the key. But let's not trust comments, let's look at the code. After we are done in this class the formatter is called:

```ruby
@set.formatter.generate(named_route, options, recall, PARAMETERIZE)
```

Here the first thing that is done is to merge the two hashes:


```ruby
def generate(name, options, path_parameters, parameterize = nil)
  constraints = path_parameters.merge(options)
```

So whether we delete the key in `recall` (which becomes `path_parameters`) the `constraints` hash will be the same.

You might be thinking "well i bet `path_parameters` is uesed somewhere else" and you would be right

It's used here: 

```ruby
parameterized_parts = extract_parameterized_parts(route, options, path_parameters, parameterize)
```

Yet once again, it's immediately merged:

```ruby
def extract_parameterized_parts(route, options, recall, parameterize = nil)
  parameterized_parts = recall.merge(options)
```

In both cases we will use the key in options if it exists. I'm pretty comfortable with this change, but I would like some more eyes. Maybe I missed a really subtle and untested use-case, for which we should certainly add some tests.